File: plugins/safe_eval/do_eval.py
import json
import resource
import sys

data = json.loads(sys.stdin.buffer.read())
resource.setrlimit(resource.RLIMIT_NPROC, (data["nproc"], data["nproc"]))
resource.setrlimit(resource.RLIMIT_AS, (data["memory"], data["memory"]))
exec(data["code"], {}, {})

File: plugins/safe_eval/do_eval.js
const posix = require("posix");
let chunks = [];

process.stdin.on("data", chunk => {
  chunks.push(chunk);
});

/** @interface */ class EvalData {
  /** @type {string} */ code;
  /** @type {number} */ nproc;
  /** @type {number} */ memory;
}

process.stdin.on("end", () => {
  /** @type {EvalData} */
  const data = JSON.parse(chunks.join());
  posix.setrlimit("as", {soft: data.memory, hard: data.memory});
  posix.setrlimit("nproc", {soft: data.nproc, hard: data.nproc});
  eval(data.code);
});
